Documentation/Evolution/RegexSyntax.md

```diff
@@ -0,0 +1,245 @@
+
+#### Quantification
+
+```
+Quantification -> QuantOperand Quantifier?
+Quantifier     -> ('*' | '+' | '?' | '{' Range '}') QuantKind?
+QuantKind      -> '?' | '+'
+```
+
+Specifies that the operand may be matched against a certain number of times.
+
+#### Groups
+
+```
+GroupStart    -> '(?' GroupKind | '('
+GroupKind     -> ':' | '|' | '>' | '=' | '!' | '*' | '<=' | '<!' | '<*'
+               | NamedGroup | MatchingOptionSeq (':' | ')')
+            
+NamedGroup    -> 'P<' GroupNameBody '>'
+               | '<' GroupNameBody '>'
+               | "'" GroupNameBody "'"
+
+GroupNameBody -> Identifier
+```
+
+Groups define a new scope within which a recursive regular expression pattern may occur. Groups have different semantics depending on how they are introduced, some may capture the nested match, some may match against the input without advancing, some may change the matching options set in the new scope, etc.
+
+#### Anchors
+
+```
+Anchor -> '^' | '$' | '\b'
+```
+
+Anchors match against a certain position in the input rather than on a particular character of the input.
+
+#### Unicode scalars
+
+
+
+#### Builtin character classes
+
+
+
+#### Custom character classes
+
+```
+CustomCharClass -> Start Set (SetOp Set)* ']'
+Start           -> '[' '^'?
+Set             -> Member+
+Member          -> CustomCharClass | !']' !SetOp (Range | Atom)
+Range           -> Atom `-` Atom
+```
+
+Custom characters classes introduce their own language, in which most regular expression metacharacters become literal
+
+
+#### Character properties
+
+### PCRE-specific syntax
+
+#### Callouts
+
+### Oniguruma-specific syntax
+
+#### Custom reference syntax
+
+#### Callout syntax
+
+#### Absent functions
+
+### ICU-specific syntax
+
+
+
+### .NET-specific syntax
+
+#### Balancing groups
+
+```
+GroupNameBody -> Identifier | Identifier? '-' Identifier
+```
+
+.NET supports the ability for a group to reference a prior group, causing the prior group to be deleted, and any intermediate matched input to become the capture of the current group.
+
+#### Character class subtraction with `-`
+
+
+
+## Syntactic differences between engines
+
+### Conflicting differences
+
+#### Character class set operations
+
+In a custom character class, some engines allow for binary set operations that take two character class inputs, and produce a new character class output. However which set operations are supported and the spellings used differ by engine.
+
+| PCRE | ICU | UTS#18 | Oniguruma | .NET | Java |
+|------|-----|--------|-----------|------|------|
+| ❌ | Intersection `&&`, Subtraction `--` | Intersection & Subtraction | Intersection `&&` | Subtraction via `-` | Intersection  `&&` |
```

Comment:
Isn't UTS\#18 "Intersection, subtraction, and symmetric difference"?

Reply:
The example they give makes it look like symmetric difference is a requirement, but the text says otherwise:

> To meet this requirement, an implementation shall supply mechanisms for union, intersection and set-difference of sets of characters within regular expression character class expressions.

So we don't have to support symmetric difference if we don't want to.

Reply:
I feel like we might as well, so I'd say we have support. I'd avoid the use of a `&` if it's not syntax, as that could be confusing.